File: StringCrypter.java
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.util.Base64;
import java.util.Formatter; // For bytesToHex, an alternative

public class StringCrypter {

    private final String key;

    public StringCrypter(String key) {
        if (key == null || key.isEmpty()) {
            throw new IllegalArgumentException("Error: Key cannot be empty.");
        }
        this.key = key;
    }

    public String encrypt(String plaintext) {
        try {
            byte[] plainBytes = plaintext.getBytes(StandardCharsets.UTF_8);
            byte[] keyBytes = this.key.getBytes(StandardCharsets.UTF_8);
            byte[] cipherBytes = new byte[plainBytes.length];
            int keyLength = keyBytes.length;

            for (int i = 0; i < plainBytes.length; i++) {
                cipherBytes[i] = (byte) (plainBytes[i] ^ keyBytes[i % keyLength]);
            }
            return Base64.getEncoder().encodeToString(cipherBytes);
        } catch (Exception e) {
            throw new RuntimeException("Encryption failed: " + e.getMessage(), e);
        }
    }

    public String decrypt(String encryptedData) {
        try {
            byte[] cipherBytes = Base64.getDecoder().decode(encryptedData);
            byte[] keyBytes = this.key.getBytes(StandardCharsets.UTF_8);
            byte[] plainBytes = new byte[cipherBytes.length];
            int keyLength = keyBytes.length;

            for (int i = 0; i < cipherBytes.length; i++) {
                plainBytes[i] = (byte) (cipherBytes[i] ^ keyBytes[i % keyLength]);
            }
            return new String(plainBytes, StandardCharsets.UTF_8);
        } catch (IllegalArgumentException e) { 
             throw new RuntimeException("Error: Decryption failed. Invalid Base64 data.", e);
        } catch (Exception e) {
            throw new RuntimeException("Error: Decryption failed: " + e.getMessage(), e);
        }
    }

    public static String generateSalt(int lengthInBytes) {
        SecureRandom random = new SecureRandom();
        byte[] saltBytes = new byte[lengthInBytes];
        random.nextBytes(saltBytes);
        return bytesToHex(saltBytes);
    }

    private static String bytesToHex(byte[] bytes) {
        StringBuilder hexString = new StringBuilder(2 * bytes.length);
        for (byte b : bytes) {
            String hex = Integer.toHexString(0xff & b);
            if (hex.length() == 1) {
                hexString.append('0');
            }
            hexString.append(hex);
        }
        return hexString.toString();
    }

    public static String processString(char type, String data) {
        String staticKey = "test_key"; 
        String userIP = ""; 
        String dynamicKey = userIP + staticKey;

        StringCrypter encryption = new StringCrypter(dynamicKey);

        try {
            if (type == 'e') {
                String salt = generateSalt(16); // 16 bytes -> 32 hex characters
                String dataWithSalt = salt + data;
                return encryption.encrypt(dataWithSalt);
            } else if (type == 'd') {
                String decryptedData = encryption.decrypt(data);
                if (decryptedData.length() < 32) { // Salt is 32 characters
                     throw new RuntimeException("Error: Decrypted data too short to contain salt.");
                }
                // String extractedSalt = decryptedData.substring(0, 32); // Salt not used after extraction in PHP
                return decryptedData.substring(32);
            } else {
                return "Error: Invalid operation type";
            }
        } catch (Exception e) {
            // Return error message similar to PHP's, but include exception message for more detail
            return "Error: " + e.getMessage(); 
        }
    }

    public static void main(String[] args) {
        String dataToEncrypt = "This is a secret message.";
        System.out.println("Original Data: " + dataToEncrypt);

        // Encrypt using Java
        String encryptedString = processString('e', dataToEncrypt);
        System.out.println("Encrypted (Java): " + encryptedString);

        // Decrypt using Java
        String decryptedString = processString('d', encryptedString);
        System.out.println("Decrypted (Java): " + decryptedString);

        // Verify
        if (dataToEncrypt.equals(decryptedString)) {
            System.out.println("Java Encryption and Decryption successful!");
        } else {
            System.out.println("Java Encryption and Decryption failed.");
            System.out.println("Expected: " + dataToEncrypt);
            System.out.println("Got: " + decryptedString);
        }
        
        System.out.println("\n--- Interoperability Notes ---");
        System.out.println("To test PHP encrypt -> Java decrypt:");
        System.out.println("1. Encrypt a string in PHP using processString('e', yourData).");
        System.out.println("2. Take the output and use it as input for Java's processString('d', phpEncryptedOutput).");
        
        System.out.println("\nTo test Java encrypt -> PHP decrypt:");
        String javaEncryptedForPHP = processString('e', "Hello from Java for PHP!");
        System.out.println("1. Java encrypted string: " + javaEncryptedForPHP);
        System.out.println("2. In PHP, use processString('d', '" + javaEncryptedForPHP + "') to decrypt.");
    }
}

File: StringCrypter.cs
using System;
using System.Text;
using System.Security.Cryptography; // For RandomNumberGenerator

public class StringCrypter
{
    private readonly byte[] _key;

    public StringCrypter(string key)
    {
        if (string.IsNullOrEmpty(key))
        {
            throw new ArgumentException("Error: Key cannot be empty.");
        }
        _key = Encoding.UTF8.GetBytes(key);
    }

    public string Encrypt(string plaintext)
    {
        try
        {
            byte[] plainBytes = Encoding.UTF8.GetBytes(plaintext);
            byte[] cipherBytes = new byte[plainBytes.Length];
            int keyLength = _key.Length;

            for (int i = 0; i < plainBytes.Length; i++)
            {
                cipherBytes[i] = (byte)(plainBytes[i] ^ _key[i % keyLength]);
            }
            return Convert.ToBase64String(cipherBytes);
        }
        catch (Exception e)
        {
            throw new ApplicationException($"Encryption failed: {e.Message}", e);
        }
    }

    public string Decrypt(string encryptedData)
    {
        try
        {
            byte[] cipherBytes = Convert.FromBase64String(encryptedData);
            byte[] plainBytes = new byte[cipherBytes.Length];
            int keyLength = _key.Length;

            for (int i = 0; i < cipherBytes.Length; i++)
            {
                plainBytes[i] = (byte)(cipherBytes[i] ^ _key[i % keyLength]);
            }
            return Encoding.UTF8.GetString(plainBytes);
        }
        catch (FormatException e)
        {
            throw new ApplicationException("Error: Decryption failed. Invalid Base64 data.", e);
        }
        catch (Exception e)
        {
            throw new ApplicationException($"Error: Decryption failed: {e.Message}", e);
        }
    }

    public static string GenerateSalt(int lengthInBytes = 16)
    {
        byte[] saltBytes = new byte[lengthInBytes];
        using (RandomNumberGenerator rng = RandomNumberGenerator.Create())
        {
            rng.GetBytes(saltBytes);
        }
        // Convert bytes to hex string
        StringBuilder hex = new StringBuilder(saltBytes.Length * 2);
        foreach (byte b in saltBytes)
        {
            hex.AppendFormat("{0:x2}", b);
        }
        return hex.ToString();
    }

    public static string ProcessString(char type, string data)
    {
        string staticKey = "test_key";
        string userIP = ""; // Kept for consistency
        string dynamicKey = userIP + staticKey;

        StringCrypter encryption = new StringCrypter(dynamicKey);

        try
        {
            if (type == 'e')
            {
                string salt = GenerateSalt(16); // 16 bytes -> 32 hex characters
                string dataWithSalt = salt + data;
                return encryption.Encrypt(dataWithSalt);
            }
            else if (type == 'd')
            {
                string decryptedData = encryption.Decrypt(data);
                if (decryptedData.Length < 32) // Salt is 32 characters
                {
                    throw new ApplicationException("Error: Decrypted data too short to contain salt.");
                }
                // string extractedSalt = decryptedData.Substring(0, 32); // Salt not used after extraction
                return decryptedData.Substring(32);
            }
            else
            {
                return "Error: Invalid operation type";
            }
        }
        catch (Exception e)
        {
            return $"Error: {e.Message}";
        }
    }

    public static void Main(string[] args)
    {
        string dataToEncrypt = "This is a secret message.";
        Console.WriteLine($"Original Data: {dataToEncrypt}");

        // Encrypt using C#
        string encryptedString = ProcessString('e', dataToEncrypt);
        Console.WriteLine($"Encrypted (C#): {encryptedString}");

        // Decrypt using C#
        string decryptedString = ProcessString('d', encryptedString);
        Console.WriteLine($"Decrypted (C#): {decryptedString}");

        // Verify
        if (dataToEncrypt.Equals(decryptedString))
        {
            Console.WriteLine("C# Encryption and Decryption successful!");
        }
        else
        {
            Console.WriteLine("C# Encryption and Decryption failed.");
            Console.WriteLine($"Expected: {dataToEncrypt}");
            Console.WriteLine($"Got: {decryptedString}");
        }

        Console.WriteLine("\n--- Interoperability Test (C# decrypts Python) ---");
        // String "Hello from Python for PHP/Java!" encrypted by Python:
        // W1dAFGgJAUwVUEtBZ1pQHUNcFhI8XAEYQARCRWlaXU0gDRoHfwIWWRVFABE8GQANVAgWBywKAhxaWGkUVGy1LNTEkSjkVKQpE
        string pythonEncrypted = "W1dAFGgJAUwVUEtBZ1pQHUNcFhI8XAEYQARCRWlaXU0gDRoHfwIWWRVFABE8GQANVAgWBywKAhxaWGkUVGy1LNTEkSjkVKQpE";
        string expectedPythonDecryption = "Hello from Python for PHP/Java!";
        Console.WriteLine($"Python Encrypted: {pythonEncrypted}");
        string decryptedFromPython = ProcessString('d', pythonEncrypted);
        Console.WriteLine($"Decrypted by C#: {decryptedFromPython}");
        if (expectedPythonDecryption.Equals(decryptedFromPython))
        {
            Console.WriteLine("C# decryption of Python string successful!");
        }
        else
        {
            Console.WriteLine("C# decryption of Python string FAILED.");
        }


        Console.WriteLine("\n--- Interoperability Test (C# encrypts for others) ---");
        string csEncryptedForOthers = ProcessString('e', "Hello from C# for other languages!");
        Console.WriteLine($"C# Encrypted for others: {csEncryptedForOthers}");
        Console.WriteLine("Take this string and try to decrypt it using process_string('d', ...) in Python or other languages.");
    }
}
